perceval/backends/core/rocketchat.py

```diff
@@ -0,0 +1,338 @@
+        """
+        return CATEGORY_MESSAGE
+
+    @staticmethod
+    def metadata_id(item):
+        """Extracts the identifier from a RocketChat item."""
+        return item["_id"]
+
+    @staticmethod
+    def metadata_updated_on(item):
+        """Extracts the update time from a RocketChat item.
+        The timestamp is extracted from 'ts' field,
+        and sliced to get the year, month and date.
+        It is then converted into a UNIX timestamp.
+        :param item: item generated by the backend
+        :returns: extracted timestamp
+        """
+        ts = item['ts'][:10]
+        ts = time.mktime(datetime.datetime.strptime(ts, "%Y-%m-%d").timetuple())
+
+        return ts
+
+    def _init_client(self, from_archive=False):
+        """Init client"""
+
+        return RocketChatClient(self.url, self.user_id, self.api_token, self.max_items, self.archive,
+                                from_archive, self.ssl_verify, roomId=self.roomId)
+
+    def __get_or_fetch_user(self, user_id, username, name):
+        if user_id in self._users:
+            return self._users[user_id]
+
+        logger.debug("User %s not found on client cache; fetching it", user_id)
+
+        user = {
+            "username": username,
+            "name": name,
+        }
+
+        self._users[user_id] = user
+        return user
+
+
+class RocketChatClient(HttpClient):
+    """RocketChat API client.
+    Client for fetching information from the RocketChat server
+    using its REST API.
+    :param api_token: key needed to use the API
+    :param max_items: maximum number of items per request
+    :param archive: an archive to store/read fetched data
+    :param from_archive: it tells whether to write/read the archive
+    :param ssl_verify: enable/disable SSL verification
+    """
+    ROOM_HISTORY = 'channels.history'
+    ROOM_INFO = 'channels.info'
+    ROOM_ID = 'roomId'
+    AUTH_TOKEN = 'X-Auth-Token'
+    USER_ID = 'X-User-Id'
+    PCOUNT = "count"
+    POLDEST = "oldest"
+
+    def __init__(self, url, user_id, api_token, max_items=MAX_ITEMS, archive=None, from_archive=False,
+                 ssl_verify=True, roomId=None):
+        super().__init__(url, archive=archive, from_archive=from_archive, ssl_verify=ssl_verify)
+        self.api_token = api_token
+        self.max_items = max_items
+        self.roomId = roomId
+        self.user_id = user_id
+
+    def room_info(self, roomId):
+        """Fetch information about a room."""
+
+        resource = self.ROOM_INFO
+
+        params = {
+            self.ROOM_ID: roomId,
+        }
+
+        response = self._fetch(resource, params)
+
+        return response
+
+    def history(self, roomId, oldest=None):
+        """Fetch the history of a room."""
+
+        resource = self.ROOM_HISTORY
+
+        params = {
+            self.ROOM_ID: roomId,
+            self.PCOUNT: self.max_items,
+            self.POLDEST: "oldest",
+        }
+
+        response = self._fetch(resource, params)
```

Comment:
How many old messages can be fetched from a channel? In Slack, the amount of messages to be retrieved is limited (that API doesn't store all historical data), do we have the same behavior in RocketChat?

Reply:
From what I read over [here](https://github.com/RocketChat/Rocket.Chat/pull/11236#issuecomment-399990958), I believe that the number of messages stored depends on the admin, and can vary for each server. Since the `oldest` parameter did not seem to work, I tried running the backend on [https://open.rocket.chat/](https://open.rocket.chat/) with `--max-items` set to 100000, and it returned 42089 messages, the last one dating to `2019-03-12` (around a year)

Reply:
Interesting, thanks @Syn3rman for the info. 

Why the param `oldest` doesn't seem to work? I understand that if you run the backend on https://open.rocket.chat/ with from-date `2019-01-01` and max-items `100000`, the last message should have date `2019-01-01`. Am I missing something?

Reply:
I made some tests and I see that `max-items/count` takes over the `oldest/from-date`. The reason should be that the client doesn't provide a mechanism to iterate over the paginated results provided by the RocketChat API (https://rocket.chat/docs/developer-guides/rest-api/offset-and-count-and-sort-info/).

An approach could be to mimic what the slack backend does (https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/slack.py#L135). Thus, the backend keeps asking for data until there is no message returned by the API. WDYT?

Reply:
> Interesting, thanks @Syn3rman for the info.
> 
> Why the param `oldest` doesn't seem to work? I understand that if you run the backend on https://open.rocket.chat/ with from-date `2019-01-01` and max-items `100000`, the last message should have date `2019-01-01`. Am I missing something?

The api seems to ignore the parameter and return the result that it would if the parameter wasn't passed. I had tried substituting the `max-items` as 100000 just to check how far back the messages were stored.

> An approach could be to mimic what the slack backend does (https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/slack.py#L135). Thus, the backend keeps asking for data until there is no message returned by the API. WDYT?

I went through the link stated above and I think that even though the API does not support the from date parameter, we could just fetch results incrementally and stop when the date is matched.
However, I think that this might be a bit slow and exhaust the rate limit, if it is very stringent. 
I tried it the below code as a standalone python script, and it appears to work:
```
fetching = True
offset = 1

while fetching:
    params = {
        "roomName": "general",
        "count": count,
        "offset": offset*count
    }
    r = requests.get(url, headers=headers,params=params)
    res = r.json()
    msg = res["messages"]
    for val in msg:
        msg_date = str_to_datetime(val["_updatedAt"])
        print(msg_date, from_date, msg_date < from_date)
        if msg_date < from_date:
            fetching = False
    offset+=1
```

Reply:
good news, thanks @Syn3rman. A similar approach has been implemented also in other backends, for instance:
- https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/meetup.py#L154
- https://github.com/chaoss/grimoirelab-perceval/blob/master/perceval/backends/core/mattermost.py#L146

Please consider that Perceval retrieves data from the oldest to the newest, take this into account when getting the data from the API (ref: https://rocket.chat/docs/developer-guides/rest-api/offset-and-count-and-sort-info/#query-parameter-information). Thanks

Reply:
Okay!
I finally figured out how to do it, I will complete this and the rate limit part in a few hours and push it :)